doc/SAI_failure_handling/SAI_failure_handling.md

```diff
@@ -0,0 +1,100 @@
+Based on the failure information, the failure handling functions in orchagent make the first attempt to address the failure.
+An ERROR_DB is also introduced to support escalation to upper layers.
+In the scenario where orchagent is unable to resolve the problem, the failure handling functions would escalate the failure to upper layers by pushing the failure into the ERROR_DB.
+<img src="Framework.png">
+
+### 2.3 Failure handling functions in Orchagent
+#### 2.3.1 Failure handling functions
+To support a failure handling logic in general while also allow each orch to have its specific logic, we include the following virtual functions in Orch
+1. `virtual task_process_status handleSaiCreateStatus(sai_api_t api, sai_status_t status, void *context = nullptr)`
+2. `virtual task_process_status handleSaiSetStatus(sai_api_t api, sai_status_t status, void *context = nullptr)`
+3. `virtual task_process_status handleSaiRemoveStatus(sai_api_t api, sai_status_t status, void *context = nullptr)`
+
+The three functions handle SAI failures in create, set, and remove operations, respectively.
+With the type of SAI API and SAI status as an input, the function could handle the failure according to the two pieces ofinformation.
+
+In the scenario where a specific logic is required in one of the Orchs, this design allows the Orch to inherit the function and include the specific login in the inherited function.
+
+The function also allows an optional input `context`, which allows passing context (e.g., object entry, attribute, etc.) into the function so that it could escalate the information to the ERROR_DB and upper layers.
+
+#### 2.3.2 Possible execution results
+1. Return `task_success` --  No crash, no retry, handled successfully. 
+
+The failure handling function should return `task_success` when the failure is properly handled without the need for another attempt (e.g., the SAI status is `SAI_STATUS_ITEM_NOT_FOUND` in remove operation).
+
+2. Return `task_failed` -- No crash, no retry, not handled successfully. 
+
+The failure handling function should return `task_failed` when the failure is unable to be handled in orchagent and another attempt is not likely to resolve the failure. In such a scenario, the function should prevent orchagent from retrying and escalate the failure to upper layers.
+
+3. Return `task_need_retry` --  No crash, retry
+
+The failure handling function should return `task_need_retry` when the failed SAI call may be resolved in a subsequent attempt.
+
+4. exit(EXIT_FAILURE) -- Crash and trigger SwSS auto-restart
+
+Some of the failures can be resolved by restarting SwSS.
+In the scenario where such failures happen, the failure handling function in orchagent should exit with `EXIT_FAILURE` and trigger SwSS auto restart.
+
+
+
+### 2.4 DB changes
+An ERROR_DB will be introduced to escalate the failures from orchagent to upper layers such as fpmsyncd.
+
+The schema of ERROR_DB is designed as follows:
```

Comment:
>schema [](start = 4, length = 6)

If you accumulate them into a table, who will delete them? otherwise it will keep growing in memory.

Reply:
Added a discussion in the design that it is necessary to avoid accumulating failures in ERROR_DB and consuming memory. To make sure all ERROR_DB entries are consumed, the failure handling should only escalate failures when the corresponding handling mechanism is available in the upper layers. Also included a possible implementation of this behavior.